CASMPET-7619/CASMPET-7620/CASMPET-7621/CASMPET-7630/CASMPET-7632: Create/fix procedures for changing iSCSI workers (#6132)
Cray-HPE/docs-csm: operations/iscsi_sbps/Managing_selective_node_personalization.md

Reviewer: This page needs more structure and explanation. It should start with an explanation of what sort of management we mean. In this case, we mean either initially specifying or later changing which worker NCNs are enabled for iSCSI. It should also highlight some specific scenarios when this is likely to be done -- install time, first upgrade from CSM 1.6 to CSM 1.7, or when adding or removing worker NCNs.

Then it should give a summary of how it works. Namely:
* All worker nodes are CONFIGURED for iSCSI
* The SBPS marshal (or whatever the right term is -- I don't know the feature as well as you do) on the workers will enable a worker as an iSCSI target if that worker has the Kubernetes label.
* When CFS performs NCN node personalization, it updates the Kubernetes labels of the worker nodes to reflect the HSM iSCSI group.

And so it can conclude with something like:
The process of initially setting these limits during an install or upgrade just involves creating the HSM group with the desired workers, and then when CFS runs later, it will do the rest.
The process of changing the limits later involves two steps:
1. Modify the HSM group
2. Re-run the iSCSI CFS layer on the workers

And then you'll have a section for how to modify the HSM group, and a section on how to run the CFS script.

Finally, the DNS issue needs to be addressed -- specifically, the manual DNS steps required when you remove a worker from being enabled after it previously had been enabled.

Reviewer: So the high level organization of the page could be something like:
```md
<table of contents with links to each subheader>

## Overview

<explain what we mean by the title of the page, and what will be found on the page>

## How it works

<explain the stuff I put in the summary in my previous comment>

## Procedures

<explain that the procedure differs based on the scenario when this is happening>

### CSM install

Something like: "When you get to the Configure Administrative Access step, if you want to use this feature, then create the HSM group at that time. Then after SAT bootprep creates the CFS configurations and they run during NCN personalization, it will take effect."

Also mention here the downsides of skipping this and later wanting to limit the worker nodes (the DNS issue).

### Upgrade from CSM 1.6 to CSM 1.7

Put the info about when they should create the HSM group, if they want to use the feature.

And in this case, the DNS info will have been added for all of the workers back when the system was running CSM 1.6. So you'll need to note here that if they do choose to use the feature, they'll need to do the manual steps for the DNS issue.

### After initial CSM 1.7 install or upgrade

Start with a preface that if this is in the context or adding or removing worker NCNs, then this should be done before the workers are added or removed.

1. Create, modify, or delete group as needed.
2. If any changes were made, run CFS script.
3. If any iSCSI-enabled workers were changed to disabled (or were removed from the system), then the DNS manual steps need to be followed.

## Manual DNS steps
```

Author: From yesterdays discussion, we decided to have command line instructions in  Managing_selective_node_personalization.md. The text explaining the feature in iscsi_sbps.md. Hope you looked at https://github.com/Cray-HPE/docs-csm/blob/release/1.7/operations/iscsi_sbps/iscsi_sbps.md#1-worker-node-personalization. 

I agree that maunal DNS steps required need to be documented and things to be taken care at compute node side also need to be documented. This we are planning to address later in a different ticket after test verification. We want the first cut of documentation to be submitted to facilitate SSI and QE teams for their installation/upgrades.

Reviewer: Yeah, I remember the discussion. Explaining the feature at a high level belongs in the main page. I just realized that it doesn't make sense to talk about the HSM group and that stuff on the main page, since that's basically an implementation detail for this feature. And the steps on this page don't really make sense unless you understand more details about how this works. So in that sense, I think it makes more sense to move some of that content back to this page. See [this comment](https://github.com/Cray-HPE/docs-csm/pull/6132/files#r2223477470) for the exact parts I thought we should move back here.

# Managing Selective Node Personalization

> Selective iSCSI worker node personalization uses an HSM group named `iscsi_worker` as the mechanism for an administrator

Reviewer: ```suggestion
## Overview

The selective iSCSI worker node personalization feature allows administrators to choose which worker NCNs are
enabled as iSCSI targets.
This feature uses an HSM group named `iscsi_worker` as the mechanism for an administrator
```

> Selective iSCSI worker node personalization uses an HSM group named `iscsi_worker` as the mechanism for an administrator
> to specify which workers should be enabled as iSCSI targets. If this group does not exist, then all worker nodes will be
> enabled as iSCSI targets. While it is technically possible to create this group and leave it empty, this will mean that
> no worker nodes will be enabled as iSCSI targets, which in turn will mean that no managed nodes will be able to boot.
> Therefore CSM generally treats it as an error if the HSM group exists but it contains no worker nodes.

Reviewer: ```suggestion
## Overview

The selective iSCSI worker node personalization feature allows administrators to specify which worker NCNs are
enabled as iSCSI targets. The mechanism by which an administrator specifies this is an HSM group named
`iscsi_worker`. The existence of this group on the system means that selective iSCSI worker node personalization
is enabled. If the group does not exist, then the feature is disabled, and all worker nodes will be
enabled as iSCSI targets. While it is technically possible to create this group and leave it empty, this will mean that
no worker nodes will be enabled as iSCSI targets, which in turn will mean that no managed nodes will be able to boot.
Therefore CSM generally treats it as an error if the HSM group exists but it contains no worker nodes.

The [Group commands](#group-commands) section goes over the commands used to create, remove, or modify the group. The [Procedures](#procedures) section discusses the different contexts in which this may be done, and any
additional steps that must be performed to effectuate the changes.

## Group commands

```

Example commands to manage selective worker node personalization for iSCSI SBPS:

(`ncn-mw#`) Create HSM group:

```bash
cray hsm groups create --label iscsi_worker --description "iscsi node personalization" --members-ids x3000c0s5b0n0
```

Example output:

> ```json

Reviewer: This is not JSON, it is TOML. If you want JSON output, you can re-run the CLI command with `--format json`

This error appears to exist in several places on this page

# Managing Selective Node Personalization

> 

Reviewer: ```suggestion

* [Overview](#overview)
* [Group commands](#group-commands)
    * [Creating the group](#creating-the-group)
    * [Adding a worker to the group](#adding-a-worker-to-the-group)
    * [Listing group members](#listing-group-members)
    * [Removing a worker from the group](#removing-a-worker-from-the-group)
    * [Deleting the group](#deleting-the-group)
* [Procedures](#procedures)
    * [CSM install](#csm-install)
    * [CSM upgrade from 1.6 to 1.7](#csm-upgrade-from-16-to-17)
    * [Adding or removing worker NCN](#adding-or-removing-worker-ncn)
    * [After initial CSM 1.7 install or upgrade](#after-initial-csm-17-install-or-upgrade)
* [Manual DNS records update](#manual-dns-records-update)

```

## Overview

The selective iSCSI worker node personalization feature allows administrators to specify which worker NCNs are enabled as

> iSCSI targets. The mechanism by which an administrator specifies this is an HSM group named `iscsi_worker`.

Reviewer: ```suggestion
iSCSI targets. The mechanism by which an administrator specifies this is the `iscsi_worker`
[component group](../hardware_state_manager/Component_Groups_and_Partitions.md#groups)
in the [Hardware State Manager (HSM)](../../glossary.md#hardware-state-manager-hsm).
```

The existence of this group on the system means that selective iSCSI worker node personalization is enabled.
If the group does not exist, then the feature is disabled, and all worker nodes will be enabled as iSCSI targets.
While it is technically possible to create this group and leave it empty, this will mean that no worker nodes will be
enabled as iSCSI targets, which in turn will mean that no managed nodes will be able to boot. Therefore CSM generally
treats it as an error if the HSM group exists but it contains no worker nodes.

The [Group commands](#group-commands) section goes over the commands used to create, remove, or modify the group.
The [Procedures](#procedures) section discusses the different contexts in which this may be done, and any additional
steps that must be performed to effectuate the changes.

## Group commands

> Example commands to manage selective worker node personalization for iSCSI SBPS:

Reviewer: ```suggestion
This section contains example commands for managing selective worker node personalization for iSCSI SBPS.
For more in-depth information on managing HSM groups in general, see
[Manage Component Groups](../hardware_state_manager/Manage_Component_Groups.md).

* [Creating the group](#creating-the-group)
* [Adding a worker to the group](#adding-a-worker-to-the-group)
* [Listing group members](#listing-group-members)
* [Removing a worker from the group](#removing-a-worker-from-the-group)
* [Deleting the group](#deleting-the-group)

### Creating the group
```

> (`ncn-mw#`) Create HSM group:

Reviewer: ```suggestion
(`ncn-mw#`) Create `iscsi_worker` group.
```

```bash

> cray hsm groups create --label iscsi_worker --description "iscsi node personalization" --members-ids x3000c0s5b0n0

Reviewer: Since it is not really a recommended case to have the group contain only a single worker node, this example should be changed to include multiple.

Author: next command following it is having adding additional worker node. Hope it addresses this ? Adding more than one xname did not go thru.

Reviewer: More than one xname works -- that's how I've been creating it usually (unless I specifically wanted just one worker). I think that is going to be the normal case, so it's probably what we should show here. It seems unlikely a customer would want to create the group with only a single xname.

Author: It does not work with two xnames:

ncn-m001:~ # cray hsm groups create --label iscsi_worker --description "iscsi node personalization" --members-ids x3000c0s7b0n0, x3001c0s35b0n0
Usage: cray hsm groups create [OPTIONS]
Try 'cray hsm groups create -h' for help.

Error: Got unexpected extra argument (x3001c0s35b0n0)

Reviewer: Did you check the HSM docs for examples? I think we added a link to the top of the group commands section specifically to get more detailed information on managing groups. Are there no examples there of creating a group with multiple xnames?

```

Example output:

```json
[[results]]
URI = "/hsm/v2/groups/iscsi_worker"
```

> HSM group `iscsi_worker` created with xname `x3000c0s5b0n0` added.

Reviewer: ```suggestion
The output does not contain the list of members in the group. If an administrator wishes to confirm that
the group was created with the correct members, see [Listing group members](#listing-group-members).
```

> 

Reviewer: ```suggestion

### Adding a worker to the group

```

> (`ncn-mw#`) Adding one more xname of worker node is as below:

Reviewer: ```suggestion
(`ncn-mw#`) Add one more worker node to the `iscsi_worker` group.

> Note: Each command may only add a single worker.
```

```bash
cray hsm groups members create --id x3000c0s18b0n0 iscsi_worker
```

Example output:

```json
[[results]]
URI = "/hsm/v2/groups/iscsi_worker/members/x3000c0s18b0n0"
```

> (`ncn-mw#`) The group members of `iscsi_worker` can be listed as below:

Reviewer: ```suggestion
### Listing group members

(`ncn-mw#`) List the members of the `iscsi_worker` group.
```

```bash
cray hsm groups members list iscsi_worker
```

Example output:

```json
ids = [ "x3000c0s5b0n0", "x3000c0s18b0n0",]
```

> (`ncn-mw#`) Deleting the worker node from `iscsi_worker` group:

Reviewer: ```suggestion
### Removing a worker from the group

(`ncn-mw#`) Remove a worker node from the `iscsi_worker` group.

> Each command may only remove a single worker.
```

```bash
cray hsm groups members delete x3000c0s18b0n0 iscsi_worker
```

Example output:

```json
code = 0
message = "deleted 1 entry"
```

> 
> (`ncn-mw#`) Checking the group after node deletion which shows empty list as below:
> 
> ```bash
> cray hsm groups members list iscsi_worker
> ids = []
> ```

Reviewer: Since an empty group is not a supported configuration, I think we don't need to include this example.

```suggestion
```

> (`ncn-mw#`) Deleting the `iscsi_worker` group:

Reviewer: ```suggestion
### Deleting the group

(`ncn-mw#`) Delete the `iscsi_worker` group.
```

```bash
cray hsm groups delete iscsi_worker
```

Example output:

```json
code = 0
message = "deleted 1 entry"
```

## Procedures

> To avail selective worker node personalization, the procedure with `iscsi_worker` group creating differs based on the
> scenario when this is happening.

Reviewer: ```suggestion
The procedures vary for enabling or disabling this feature, or modifying the set of worker nodes designated as iSCSI
targets. It depends on whether this is being done during a fresh install of CSM, an upgrade from CSM 1.6
to CSM 1.7, or on an operational CSM 1.7+ system.

* [CSM install](#csm-install)
* [CSM upgrade from 1.6 to 1.7](#csm-upgrade-from-16-to-17)
* [Adding or removing worker NCN](#adding-or-removing-worker-ncn)
* [After initial CSM 1.7 install or upgrade](#after-initial-csm-17-install-or-upgrade)
```

### CSM install

> The `iscsi_worker` HSM group needs to be created just before the `Configure management nodes with CFS` step in

Reviewer: ```suggestion
During a CSM install, if the administrator wants to enable this feature, they should
[create the `iscsi_worker` HSM group](#creating-the-group) as part of
```

> [configure administrative access](https://github.com/Cray-HPE/docs-csm/blob/release/1.7/install/configure_administrative_access.md).

Reviewer: This needs to be a relative link, and `Configure Administrative Access` should be capitalized in the link text.

> This will take into effect when SAT bootprep creates the CFS configurations and run NCN personalization.

Reviewer: ```suggestion
The feature will take effect when
[Management Node Personalization](../configuration_management/Management_Node_Personalization.md)
runs on the NCNs, after [SAT Bootprep](../system_admin_toolkit/usage/SAT_Bootprep.md) has run.
```

If creating `iscsi_worker` group is skipped and if admin wants to create post CSM install, then it requires additional

> things to be taken care like updating DNS SRV and A records manually.

Reviewer: ```suggestion
If the group is not created at this time, and an administrator wishes to enable this feature post-install, additional
manual steps are required. This is also true if an administrator creates the group during the install, but later
wants to disable iSCSI on some of the workers where it had previously been enabled.

For details, see [After initial CSM 1.7 install or upgrade](#after-initial-csm-17-install-or-upgrade).
```

> ### CSM Upgrade from 1.6 to 1.7

Reviewer: ```suggestion
### CSM upgrade from 1.6 to 1.7
```

> The `iscsi_worker` HSM group needs to be created just before the `Management node rollout` mentioned in [upgrade CSM and additional products with IUF](../iuf/workflows/upgrade_csm_and_additional_products_with_iuf.md)

Reviewer: ```suggestion
During an upgrade from CSM 1.6 to CSM 1.7, if the administrator wants to enable this feature, they should
[create the `iscsi_worker` HSM group](#creating-the-group) just before the
[Management Rollout](iuf/workflows/management_rollout.md) step of
[Upgrade CSM and additional products with IUF](../iuf/workflows/upgrade_csm_and_additional_products_with_iuf.md).
```

> The `iscsi_worker` HSM group needs to be created just before the `Management node rollout` mentioned in [upgrade CSM and additional products with IUF](../iuf/workflows/upgrade_csm_and_additional_products_with_iuf.md)
> 
> In this case, the DNS information will have been added for all of the worker nodes back when the system was running
> CSM 1.6. So if selective worker node personalization feature is to be used, DNS SRV and A records have to be updated
> manually.

Reviewer: ```suggestion
During an upgrade from CSM 1.6 to CSM 1.7, if the administrator does not wish to enable this
feature, no special steps need to be taken.

If the administrator does wish to enable this feature, then the following steps need to be
followed at the beginning of
[2. Execute the IUF `management-nodes-rollout` stage](../iuf/workflows/management_rollout.md#2-execute-the-iuf-management-nodes-rollout-stage).

1. Create the `iscsi_worker` HSM group.

    See [Creating the group](#creating-the-group).

1. Perform manual DNS update.

    This step is required any time a worker node is changed from being enabled for iSCSI to
    being disabled for iSCSI. In CSM 1.6, all worker nodes were enabled for iSCSI. Therefore,
    this step will always be required when first enabling this feature on a system that
    previously was running CSM 1.6.

    See [Manual DNS records update](#manual-dns-records-update).
```

> ### After initial CSM 1.7 install or upgrade

Reviewer: ```suggestion
### Adding or removing worker NCN

When [adding or removing a worker NCN](../node_management/Add_Remove_Replace_NCNs/Add_Remove_Replace_NCNs.md),
the administrator may also wish to create, modify, or delete the `iscsi_worker` HSM group, in order to have the
desired set of worker NCNs enabled as iSCSI targets. If this is the case, then this should be done **before**
adding or removing the worker NCN.
For the procedure itself, see [After initial CSM 1.7 install or upgrade](#after-initial-csm-17-install-or-upgrade).

### After initial CSM 1.7 install or upgrade
```

> In the scenario where adding or removing worker NCNs, then `iscsi_worker` HSM group needs to be created before the worker nodes are added or removed.

Reviewer: ```suggestion
> If this is being done in the context of adding or removing a worker NCN, these steps should be done
> **before** carrying out that procedure.
```

In the scenario where `iscsi_worker` group is not created during CSM install/upgrade, then it can be created post CSM install/upgrade and also in the scenario where any modifications to the group is to be done post install/upgrade, like
adding/removing worker nodes to/from the group, then it is required to re-run the iSCSI CFS layer on worker nodes and
update DNS SRV and A records manually.

Re-running the iSCSI CFS layer can be done using the below script:

```bash
/usr/share/doc/csm/scripts/operations/configuration/refresh_worker_iscsi_config.py
```

Also, The latest CSM documentation RPM must be installed on the node where the procedure is being performed.

> See [Check for latest documentation](../../update_product_stream/README.md#check-for-latest-documentation)

Reviewer: ```suggestion
See [Check for latest documentation](../../update_product_stream/README.md#check-for-latest-documentation)

## Manual DNS records update

One of the things that the iSCSI Ansible playbook does is to add DNS entries for the workers that are
enabled for iSCSI. However, this playbook does not have logic to remove these entries, in the case where
a worker had previously been enabled for iSCSI but now is not. In this case, the administrator must manually remove
these entries.

These entries are added by the `csm.sbps.dns_srv_records` Ansible role in the `csm-config-management`
repository in the [Version Control Service (VCS)](../../glossary.md#version-control-service-vcs).
See that role to determine which entries must be manually removed.
```

> In the scenario where `iscsi_worker` group is not created during CSM install/upgrade, then it can be created post CSM install/upgrade and also in the scenario where any modifications to the group is to be done post install/upgrade, like
> adding/removing worker nodes to/from the group, then it is required to re-run the iSCSI CFS layer on worker nodes and
> update DNS SRV and A records manually.
> 
> Re-running the iSCSI CFS layer can be done using the below script:
> 
> ```bash
> /usr/share/doc/csm/scripts/operations/configuration/refresh_worker_iscsi_config.py
> ```
> 
> Also, The latest CSM documentation RPM must be installed on the node where the procedure is being performed.
> See [Check for latest documentation](../../update_product_stream/README.md#check-for-latest-documentation)

Reviewer: ```suggestion
During normal system operation, this feature can be enabled or disabled, or the set of worker nodes being
enabled for iSCSI can be changed. To do this, use the following procedure:

1. Create, modify, or delete the `iscsi_worker` HSM group.

    * If this feature is currently disabled (meaning that all worker nodes are enabled as iSCSI targets),
      then it can be enabled by creating the HSM group and specifying which worker nodes should
      be enabled for iSCSI.

      See [Creating the group](#creating-the-group).

    * If this feature is currently enabled (meaning that the `iscsi_worker` HSM group exists), then
      there are two options:
      
        * The feature can be disabled (meaning that all worker nodes will be enabled as iSCSI targets)
          by deleting the group. See [Deleting the group](#deleting-the-group).

        * The set of worker NCNs being enabled for iSCSI can be modified by changing the membership of
          the group. This can be done by adding workers to it, removing workers from it, or a combination
          of both. See [Adding a worker to the group](#adding-a-worker-to-the-group) and
          [Removing a worker from the group](#removing-a-worker-from-the-group).

1. Instruct the [Configuration Framework Service (CFS)](../../glossary.md#configuration-framework-service-cfs) to
   update the iSCSI configuration.

    > The latest CSM documentation RPM must be installed on the node where this step is being performed.
    > See [Check for latest documentation](../../update_product_stream/README.md#check-for-latest-documentation)

    (`ncn-mw#`) The `refresh_worker_iscsi_config.py` script is provided to modify the CFS state of the worker NCNs to
    force them to re-run the iSCSI configuration layer. It is run without arguments.

    ```bash
    /usr/share/doc/csm/scripts/operations/configuration/refresh_worker_iscsi_config.py
    ```

1. Manually update DNS records, if needed.

    If the changes made in the first step result in a worker being disabled as an iSCSI target when previously
    it had been enabled as an iSCSI target, then an additional manual procedure is required.
    See [Manual DNS records update](#manual-dns-records-update).

## Manual DNS records update

One of the things that the iSCSI Ansible playbook does is to add DNS entries for the workers that are
enabled for iSCSI. However, this playbook does not have logic to remove these entries, in the case where
a worker had previously been enabled for iSCSI but now is not. In this case, the administrator must manually remove
these entries.

These entries are added by the `csm.sbps.dns_srv_records` Ansible role in the `csm-config-management`
repository in the [Version Control Service (VCS)](../../glossary.md#version-control-service-vcs).
See that role to determine which entries must be manually removed.
```

# Managing Selective Node Personalization

* [Overview](#overview)
* [Group commands](#group-commands)
    * [Creating the group](#creating-the-group)
    * [Adding a worker to the group](#adding-a-worker-to-the-group)
    * [Listing group members](#listing-group-members)
    * [Removing a worker from the group](#removing-a-worker-from-the-group)
    * [Deleting the group](#deleting-the-group)
* [Procedures](#procedures)
    * [CSM install](#csm-install)
    * [CSM upgrade from 1.6 to 1.7](#csm-upgrade-from-16-to-17)

>     * [Adding or removing worker NCN](#adding-or-removing-worker-ncn)

Reviewer: ```suggestion
    * [Adding or removing worker NCN](#adding-or-removing-worker-ncn)
        * [Adding worker NCN when `iscsi_sbps` group does not exist](#adding-worker-ncn-when-iscsisbps-group-does-not-exist)
        * [Adding worker NCN when `iscsi_sbps` group exists](#adding-worker-ncn-when-iscsisbps-group-exists)
        * [Removing worker NCN when `iscsi_sbps` group does not exist](#removing-worker-ncn-when-iscsisbps-group-does-not-exist)
        * [Removing worker NCN when `iscsi_sbps` group exists](#removing-worker-ncn-when-iscsisbps-group-exists)
```

The existence of this group on the system means that selective iSCSI worker node personalization is enabled.
If the group does not exist, then the feature is disabled, and all worker nodes will be enabled as iSCSI targets.
While it is technically possible to create this group and leave it empty, this will mean that no worker nodes will be
enabled as iSCSI targets, which in turn will mean that no managed nodes will be able to boot. Therefore CSM generally
treats it as an error if the HSM group exists but it contains no worker nodes.

The [Group commands](#group-commands) section goes over the commands used to create, remove, or modify the group.
The [Procedures](#procedures) section discusses the different contexts in which this may be done, and any additional
steps that must be performed to effectuate the changes.

## Group commands

Example commands to manage selective worker node personalization for iSCSI SBPS:

This section contains example commands for managing selective worker node personalization for iSCSI SBPS.

> For more in-depth information on managing HSM groups in general, see [Manage Component Groups](../hardware_state_manager/Manage_Component_Groups.md)

Reviewer: ```suggestion
For more in-depth information on managing HSM groups in general, see
[Manage Component Groups](../hardware_state_manager/Manage_Component_Groups.md).
```

* [Creating the group](#creating-the-group)
* [Adding a worker to the group](#adding-a-worker-to-the-group)
* [Listing group members](#listing-group-members)
* [Removing a worker from the group](#removing-a-worker-from-the-group)
* [Deleting the group](#deleting-the-group)

### Creating the group

> (`ncn-mw#`) Create `iscsi_worker` group

Reviewer: ```suggestion
(`ncn-mw#`) Create the `iscsi_worker` group.
```

```bash
cray hsm groups create --label iscsi_worker --description "iscsi node personalization" --members-ids x3000c0s5b0n0
```

Example output:

```toml
[[results]]
URI = "/hsm/v2/groups/iscsi_worker"
```

The output does not contain the list of members in the group. If an administrator wishes to confirm that the group was created with the correct members, see [Listing group members](#listing-group-members).

### Adding a worker to the group

(`ncn-mw#`) Add one more worker node to the `iscsi_worker` group.

> Note: Each command may only add a single worker.

Reviewer: ```suggestion
> Note: Each command may only add a single worker.
```

```bash
cray hsm groups members create --id x3000c0s18b0n0 iscsi_worker
```

Example output:

```json
[[results]]
URI = "/hsm/v2/groups/iscsi_worker/members/x3000c0s18b0n0"
```

### Listing group members

(`ncn-mw#`) List the members of the `iscsi_worker` group.

```bash
cray hsm groups members list iscsi_worker
```

Example output:

```json
ids = [ "x3000c0s5b0n0", "x3000c0s18b0n0",]
```

### Removing a worker from the group

(`ncn-mw#`) Remove a worker node from the `iscsi_worker` group.

> Each command may only remove a single worker.

Reviewer: ```suggestion
> Each command may only remove a single worker.
```

```bash
cray hsm groups members delete x3000c0s18b0n0 iscsi_worker
```

Example output:

```json
code = 0
message = "deleted 1 entry"
```

### Deleting the group

> (`ncn-mw#`) Delete the `iscsi_worker` group:

Reviewer: ```suggestion
(`ncn-mw#`) Delete the `iscsi_worker` group.
```

```bash
cray hsm groups delete iscsi_worker
```

Example output:

```json
code = 0
message = "deleted 1 entry"
```

## Procedures

> The procedures vary for enabling or disabling this feature, or modifying the set of worker nodes designated as iSCSI

Reviewer: ```suggestion
The procedures vary for enabling or disabling this feature, or modifying the set of worker nodes that are enabled as iSCSI
```

targets. It depends on whether this is being done during a fresh install of CSM, an upgrade from CSM 1.6 to CSM 1.7,

> or on an operational CSM 1.7+ system.

Reviewer: ```suggestion
as part of adding or removing a worker NCN, or on an operational CSM 1.7+ system.
```

* [CSM install](#csm-install)
* [CSM upgrade from 1.6 to 1.7](#csm-upgrade-from-16-to-17)

> * [Adding or removing worker NCN](#adding-or-removing-worker-ncn)

Reviewer: ```suggestion
* [Adding or removing worker NCN](#adding-or-removing-worker-ncn)
    * [Adding worker NCN when `iscsi_sbps` group does not exist](#adding-worker-ncn-when-iscsisbps-group-does-not-exist)
    * [Adding worker NCN when `iscsi_sbps` group exists](#adding-worker-ncn-when-iscsisbps-group-exists)
    * [Removing worker NCN when `iscsi_sbps` group does not exist](#removing-worker-ncn-when-iscsisbps-group-does-not-exist)
    * [Removing worker NCN when `iscsi_sbps` group exists](#removing-worker-ncn-when-iscsisbps-group-exists)
```

manual steps are required. This is also true if an administrator creates the group during the install, but later wants
to disable iSCSI on some of the workers where it had previously been enabled.

For details, see [After initial CSM 1.7 install or upgrade](#after-initial-csm-17-install-or-upgrade).

### CSM upgrade from 1.6 to 1.7

During an upgrade from CSM 1.6 to CSM 1.7, if the administrator does not wish to enable this feature, no special steps need to be taken. If the administrator does wish to enable this feature, then the following steps need to be followed at the
beginning of [2. Execute the IUF `management-nodes-rollout` stage](../iuf/workflows/management_rollout.md#2-execute-the-iuf-management-nodes-rollout-stage).

1. Create the `iscsi_worker` HSM group.

   See [Creating the group](#creating-the-group).

1. Perform manual DNS update.

This step is required any time a worker node is changed from being enabled for iSCSI to being disabled for iSCSI.
In CSM 1.6, all worker nodes were enabled for iSCSI. Therefore, this step will always be required when first enabling this feature on a system that previously was running CSM 1.6.

> See [Manual DNS records update](#manual-dns-records-update).

Reviewer: ```suggestion
   This step is required any time a worker node is changed from being enabled for iSCSI to being disabled for iSCSI.
   In CSM 1.6, all worker nodes were enabled for iSCSI. Therefore, this step will always be required when first enabling
   this feature on a system that previously was running CSM 1.6.

   See [Manual DNS records update](#manual-dns-records-update).
```

### Adding or removing worker NCN

When [adding or removing a worker NCN](../node_management/Add_Remove_Replace_NCNs/Add_Remove_Replace_NCNs.md), the
administrator may also wish to create, modify, or delete the `iscsi_worker` HSM group, in order to have the desired
set of worker NCNs enabled as iSCSI targets. If this is the case, then this should be done **before** adding or
removing the worker NCN.

> For the procedure itself, see [After initial CSM 1.7 install or upgrade](#after-initial-csm-17-install-or-upgrade).

Reviewer: ````suggestion
When [adding or removing a worker NCN](../node_management/Add_Remove_Replace_NCNs/Add_Remove_Replace_NCNs.md),
additional steps may be required. **Before** performing the add or remove procedure for the worker NCN,
read the section that applies to the current operation and whether or not the `iscsi_sbps` group currently exists:

> If unsure whether or not the group exists, try [Listing the group members](#listing-group-members).
> If it succeeds, the group exists (and its member list will be shown). If the group does not exist,
> the command will fail with an error that it cannot find the group.

* [Adding worker NCN when `iscsi_sbps` group does not exist](#adding-worker-ncn-when-iscsisbps-group-does-not-exist)
* [Adding worker NCN when `iscsi_sbps` group exists](#adding-worker-ncn-when-iscsisbps-group-exists)
* [Removing worker NCN when `iscsi_sbps` group does not exist](#removing-worker-ncn-when-iscsisbps-group-does-not-exist)
* [Removing worker NCN when `iscsi_sbps` group exists](#removing-worker-ncn-when-iscsisbps-group-exists)

#### Adding worker NCN when `iscsi_sbps` group does not exist

This means that all current worker NCNs are enabled as iSCSI targets.
If the new worker NCN being added should also be enabled as an iSCSI target, then no additional steps are required.
Skip the rest of this document and perform the procedure to add the worker NCN to the system.

Otherwise (if the new worker NCN should NOT be configured as an iSCSI target), perform the following steps **before**
following the add NCN procedure:

1. (`ncn-mw#`) Get the xnames of all of the worker NCNs currently in the system.

    ```bash
    WORKER_XNAMES=$(cray hsm state components list --role Management --subrole Worker --type Node --format json \
                    | jq -r .Components[].ID | sort -u | tr '\n' , | sed 's/,$//')
    echo ${WORKER_XNAMES}
    ```

    Example output:

    ```text
    x3000c0s7b0n0,x3000c0s9b0n0,x3000c0s38b0n0,x3000c0s11b0n0
    ```

1. Create the `iscsi_sbps` group.
 
    The `iscsi_sbps` group needs to be created and contain the xnames of all of the worker NCNs except the one being added
    (the same list obtained in the previous step). See [Creating the group](#creating-the-group).

1. Skip the rest of this document and perform the procedure to add the worker NCN to the system. When the new worker performs
   [Management Node Personalization](../configuration_management/Management_Node_Personalization.md), it will not be
   enabled as an iSCSI target.

#### Adding worker NCN when `iscsi_sbps` group exists

If the new worker NCN being added should not be configured as an iSCSI target, then no additional steps are required.
Skip the rest of this document and perform the procedure to add the worker NCN to the system.

Otherwise (if the new worker NCN should be configured as an iSCSI target), perform the following steps:

1. (`ncn-mw#`) Get the xnames of all of the worker NCNs currently in the system.

    ```bash
    echo $(cray hsm state components list --role Management --subrole Worker --type Node --format json \
           | jq -r .Components[].ID | sort -u)
    ```

    Example output:

    ```text
    x3000c0s11b0n0 x3000c0s38b0n0 x3000c0s7b0n0 x3000c0s9b0n0
    ```

1. Proceed with the add worker NCN procedure until the xname for the new worker NCN exists in HSM.

    This can be checked by running the same command as in the previous step, and comparing the output.
    As soon as a new xname shows up in the list, that is the xname of the worker being added.

1. Add the xname of the new worker node to the `iscsi_sbps` group.

    See [Adding a worker to the group](#adding-a-worker-to-the-group).

1. Skip the rest of this document and perform the rest of the procedure to add the worker NCN to the system. When new worker performs
   [Management Node Personalization](../configuration_management/Management_Node_Personalization.md), it will be
   enabled as an iSCSI target.

#### Removing worker NCN when `iscsi_sbps` group does not exist

This means that all worker NCNs are configured as iSCSI targets.
In this scenario, the remove NCN procedure can be performed as usual. After it is done, the
administrator must manually update DNS records. See [Manual DNS records update](#manual-dns-records-update).

#### Removing worker NCN when `iscsi_sbps` group exists

1. Check if the worker being removed is not a member of the `iscsi_sbps` group.

    See [Listing group members](#listing-group-members) for how to see the current members of the group.

1. If the worker being removed is not a member of the group, then skip the rest of this document and perform the procedure to
   remove the worker NCN from the system. No additional steps are required in this case. Otherwise, if the worker being removed
   is a member of the group, then proceed with the remaining steps in this section.

1. Remove the worker being removed from the group.

    See [Removing a worker from the group](#removing-a-worker-from-the-group).

1. Instruct the [Configuration Framework Service (CFS)](../../glossary.md#configuration-framework-service-cfs) to update
   the iSCSI configuration.

   > The latest CSM documentation RPM must be installed on the node where this step is being performed.
   > See [Check for latest documentation](../../update_product_stream/README.md#check-for-latest-documentation).

   (`ncn-mw#`) The `refresh_worker_iscsi_config.py` script is provided to modify the CFS state of the worker NCNs to
   force them to re-run the iSCSI configuration layer. It is run without arguments.

   ```bash
   /usr/share/doc/csm/scripts/operations/configuration/refresh_worker_iscsi_config.py
   ```

1. Manually update the DNS records.

    See [Manual DNS records update](#manual-dns-records-update).

1. Skip the rest of this document and perform the procedure to remove the worker NCN from the system.
````

### After initial CSM 1.7 install or upgrade

If this is being done in the context of adding or removing a worker NCN, these steps should be done **before**

>  carrying out that procedure.

Reviewer: ```suggestion
**IMPORTANT**: If this is being done in the context of adding or removing a worker NCN, see
[Adding or removing worker NCN](#adding-or-removing-worker-ncn) instead.
```

During normal system operation, this feature can be enabled or disabled, or the set of worker nodes being enabled
for iSCSI can be changed. To do this, use the following procedure:

1. Create, modify, or delete the `iscsi_worker` HSM group.

    * If this feature is currently disabled (meaning that all worker nodes are enabled as iSCSI targets), then it can be
      enabled by creating the HSM group and specifying which worker nodes should be enabled for iSCSI.

>       See [Creating the group](#creating-the-group).

Reviewer: ```suggestion

      See [Creating the group](#creating-the-group).
```

    * If this feature is currently enabled (meaning that the `iscsi_worker` HSM group exists),
      then there are two options:

> 
> ```text
>     * The feature can be disabled (meaning that all worker nodes will be enabled as iSCSI targets) by deleting
>       the group. See [Deleting the group](#deleting-the-group).
>           
>     * The set of worker NCNs being enabled for iSCSI can be modified by changing the membership of the group.
>       This can be done by adding workers to it, removing workers from it, or a combination of both.
>       See [Adding a worker to the group](#adding-a-worker-to-the-group) and [Removing a worker from the group](#removing-a-worker-from-the-group).
> ```

Reviewer: ```suggestion
        * The feature can be disabled (meaning that all worker nodes will be enabled as iSCSI targets) by deleting
          the group. See [Deleting the group](#deleting-the-group).
        * The set of worker NCNs being enabled for iSCSI can be modified by changing the membership of the group.
          This can be done by adding workers to it, removing workers from it, or a combination of both.
          See [Adding a worker to the group](#adding-a-worker-to-the-group) and
          [Removing a worker from the group](#removing-a-worker-from-the-group).
```

1. Instruct the [Configuration Framework Service (CFS)](../../glossary.md#configuration-framework-service-cfs) to update
   the iSCSI configuration.

   The latest CSM documentation RPM must be installed on the node where this step is being performed.

>    See [Check for latest documentation](../../update_product_stream/README.md#check-for-latest-documentation)

Reviewer: ```suggestion
   > The latest CSM documentation RPM must be installed on the node where this step is being performed.
   > See [Check for latest documentation](../../update_product_stream/README.md#check-for-latest-documentation).
```

   (`ncn-mw#`) The `refresh_worker_iscsi_config.py` script is provided to modify the CFS state of the worker NCNs to
   force them to re-run the iSCSI configuration layer. It is run without arguments.

   ```bash
   /usr/share/doc/csm/scripts/operations/configuration/refresh_worker_iscsi_config.py
   ```

> 

Reviewer: ```suggestion

1. Manually update DNS records, if needed.

    If the changes made in the first step result in a worker being disabled as an iSCSI target when previously
    it had been enabled as an iSCSI target, then an additional manual procedure is required.
    See [Manual DNS records update](#manual-dns-records-update).

```

## Manual DNS records update

One of the things that the iSCSI Ansible playbook does is to add DNS entries for the workers that are enabled for iSCSI.
However, this playbook does not have logic to remove these entries, in the case where a worker had previously been enabled for iSCSI but now is not. In this case, the administrator must manually remove these entries.

These entries are added by the `csm.sbps.dns_srv_records` Ansible role in the `csm-config-management` repository in the [Version Control Service (VCS)](../../glossary.md#version-control-service-vcs).

> 

Reviewer: ```suggestion
```

    * [Listing group members](#listing-group-members)
    * [Removing a worker from the group](#removing-a-worker-from-the-group)
    * [Deleting the group](#deleting-the-group)
* [Procedures](#procedures)
    * [CSM install](#csm-install)
    * [CSM upgrade from 1.6 to 1.7](#csm-upgrade-from-16-to-17)
    * [Adding or removing worker NCN](#adding-or-removing-worker-ncn)
        * [Adding worker NCN when `iscsi_sbps` group does not exist](#adding-worker-ncn-when-iscsi_sbps-group-does-not-exist)
        * [Adding worker NCN when `iscsi_sbps` group exists](#adding-worker-ncn-when-iscsi_sbps-group-exists)
        * [Removing worker NCN when `iscsi_sbps` group does not exist](#removing-worker-ncn-when-iscsi_sbps-group-does-not-exist)
        * [Removing worker NCN when `iscsi_sbps` group exists](#removing-worker-ncn-when-iscsi_sbps-group-exists)
    * [After initial CSM 1.7 install or upgrade](#after-initial-csm-17-install-or-upgrade)
* [Manual DNS records update](#manual-dns-records-update)

## Overview

The selective iSCSI worker node personalization feature allows administrators to specify which worker NCNs are enabled as

> iSCSI targets. The mechanism by which an administrator specifies this is the `iscsi_worker` [component group](../hardware_state_manager/Component_Groups_and_Partitions.md#groups) in the [Hardware State Manager (HSM)](../../glossary.md#hardware-state-manager-hsm).

Reviewer: Break up long line.
```suggestion
iSCSI targets. The mechanism by which an administrator specifies this is the `iscsi_worker`
[component group](../hardware_state_manager/Component_Groups_and_Partitions.md#groups) in the
[Hardware State Manager (HSM)](../../glossary.md#hardware-state-manager-hsm).
```

* [Removing worker NCN when `iscsi_sbps` group does not exist](#removing-worker-ncn-when-iscsi_sbps-group-does-not-exist)
* [Removing worker NCN when `iscsi_sbps` group exists](#removing-worker-ncn-when-iscsi_sbps-group-exists)

#### Adding worker NCN when `iscsi_sbps` group does not exist

This means that all current worker NCNs are enabled as iSCSI targets.
If the new worker NCN being added should also be enabled as an iSCSI target, then no additional steps are required.
Skip the rest of this document and perform the procedure to add the worker NCN to the system.

Otherwise (if the new worker NCN should NOT be configured as an iSCSI target), perform the following steps **before**
following the add NCN procedure:

1. (`ncn-mw#`) Get the xnames of all of the worker NCNs currently in the system.

    ```bash
    WORKER_XNAMES=$(cray hsm state components list --role Management --subrole Worker --type Node --format json \
                    | jq -r .Components[].ID | sort -u | tr '\n' , | sed 's/,$//')
    echo ${WORKER_XNAMES}
    ```

    Example output:

    ```text
    x3000c0s7b0n0,x3000c0s9b0n0,x3000c0s38b0n0,x3000c0s11b0n0
    ```

1. Create the `iscsi_sbps` group.

>     The `iscsi_sbps` group needs to be created and contain the xnames of all of the worker NCNs except the one being added

Reviewer: ```suggestion

    The `iscsi_sbps` group needs to be created and contain the xnames of all of the worker NCNs except the one being added
```

iSCSI targets. The mechanism by which an administrator specifies this is an HSM group named `iscsi_worker` [component group](../hardware_state_manager/Component_Groups_and_Partitions.md#groups) in the [Hardware State Manager (HSM)](../../glossary.md#hardware-state-manager-hsm).
The existence of this group on the system means that selective iSCSI worker node personalization is enabled.
If the group does not exist, then the feature is disabled, and all worker nodes will be enabled as iSCSI targets.
While it is technically possible to create this group and leave it empty, this will mean that no worker nodes will be
enabled as iSCSI targets, which in turn will mean that no managed nodes will be able to boot. Therefore CSM generally
treats it as an error if the HSM group exists but it contains no worker nodes.

The [Group commands](#group-commands) section goes over the commands used to create, remove, or modify the group.
The [Procedures](#procedures) section discusses the different contexts in which this may be done, and any additional
steps that must be performed to effectuate the changes.

## Group commands

> 
> Example commands to manage selective worker node personalization for iSCSI SBPS:

Reviewer: ```suggestion
```

    * [Removing worker NCN when `iscsi_sbps` group exists](#removing-worker-ncn-when-iscsi_sbps-group-exists)
* [After initial CSM 1.7 install or upgrade](#after-initial-csm-17-install-or-upgrade)

### CSM install

During a CSM install, if the administrator wants to enable this feature, they should [create the `iscsi_worker` HSM group](#creating-the-group) as part of [Configure Administrative Access](../../install/configure_administrative_access.md)

The feature will take effect when [Management Node Personalization](../configuration_management/Management_Node_Personalization.md) runs on the NCNs, after [SAT Bootprep](../system_admin_toolkit/usage/SAT_Bootprep.md) has run.

If the group is not created at this time, and an administrator wishes to enable this feature post-install, additional
manual steps are required. This is also true if an administrator creates the group during the install, but later wants
to disable iSCSI on some of the workers where it had previously been enabled.

For details, see [After initial CSM 1.7 install or upgrade](#after-initial-csm-17-install-or-upgrade).

### CSM upgrade from 1.6 to 1.7

> During an upgrade from CSM 1.6 to CSM 1.7, if the administrator does not wish to enable this feature, no special steps need to be taken. If the administrator does wish to enable this feature, then the following steps need to be followed at the

Reviewer: ```suggestion
During an upgrade from CSM 1.6 to CSM 1.7, if the administrator does not wish to enable this feature, no special steps
need to be taken. If the administrator does wish to enable this feature, then the following steps need to be followed at the
```

* [Removing worker NCN when `iscsi_sbps` group does not exist](#removing-worker-ncn-when-iscsi_sbps-group-does-not-exist)
* [Removing worker NCN when `iscsi_sbps` group exists](#removing-worker-ncn-when-iscsi_sbps-group-exists)

#### Adding worker NCN when `iscsi_sbps` group does not exist

This means that all current worker NCNs are enabled as iSCSI targets.
If the new worker NCN being added should also be enabled as an iSCSI target, then no additional steps are required.
Skip the rest of this document and perform the procedure to add the worker NCN to the system.

Otherwise (if the new worker NCN should NOT be configured as an iSCSI target), perform the following steps **before**
following the add NCN procedure:

1. (`ncn-mw#`) Get the xnames of all of the worker NCNs currently in the system.

    ```bash
    WORKER_XNAMES=$(cray hsm state components list --role Management --subrole Worker --type Node --format json \
                    | jq -r .Components[].ID | sort -u | tr '\n' , | sed 's/,$//')
    echo ${WORKER_XNAMES}
    ```

    Example output:

    ```text
    x3000c0s7b0n0,x3000c0s9b0n0,x3000c0s38b0n0,x3000c0s11b0n0
    ```

1. Create the `iscsi_sbps` group.

>  

Reviewer: ```suggestion

```

The [Group commands](#group-commands) section goes over the commands used to create, remove, or modify the group.
The [Procedures](#procedures) section discusses the different contexts in which this may be done, and any additional
steps that must be performed to effectuate the changes.

## Group commands

This section contains example commands for managing selective worker node personalization for iSCSI SBPS.

For more in-depth information on managing HSM groups in general, see
[Manage Component Groups](../hardware_state_manager/Manage_Component_Groups.md).

* [Creating the group](#creating-the-group)
* [Adding a worker to the group](#adding-a-worker-to-the-group)
* [Listing group members](#listing-group-members)
* [Removing a worker from the group](#removing-a-worker-from-the-group)
* [Deleting the group](#deleting-the-group)

### Creating the group

(`ncn-mw#`) Create the `iscsi_worker` group.

```bash
cray hsm groups create --label iscsi_worker --description "iscsi node personalization" --members-ids x3000c0s5b0n0,x3001c0s35b0n0
```

Example output:

```toml
[[results]]
URI = "/hsm/v2/groups/iscsi_worker"
```

> The output does not contain the list of members in the group. If an administrator wishes to confirm that the group was created with the correct members, see [Listing group members](#listing-group-members).

Reviewer: Break up long line
```suggestion
The output does not contain the list of members in the group. If an administrator wishes to confirm that the group was
created with the correct members, see [Listing group members](#listing-group-members).
```

```bash
cray hsm groups delete iscsi_worker
```

Example output:

```toml
code = 0
message = "deleted 1 entry"
```

## Procedures

The procedures vary for enabling or disabling this feature, or modifying the set of worker nodes that are enabled as iSCSI
targets. It depends on whether this is being done during a fresh install of CSM, an upgrade from CSM 1.6 to CSM 1.7,
as part of adding or removing a worker NCN, or on an operational CSM 1.7+ system.

* [CSM install](#csm-install)
* [CSM upgrade from 1.6 to 1.7](#csm-upgrade-from-16-to-17)
* [Adding or removing worker NCN](#adding-or-removing-worker-ncn)
    * [Adding worker NCN when `iscsi_sbps` group does not exist](#adding-worker-ncn-when-iscsi_sbps-group-does-not-exist)
    * [Adding worker NCN when `iscsi_sbps` group exists](#adding-worker-ncn-when-iscsi_sbps-group-exists)
    * [Removing worker NCN when `iscsi_sbps` group does not exist](#removing-worker-ncn-when-iscsi_sbps-group-does-not-exist)
    * [Removing worker NCN when `iscsi_sbps` group exists](#removing-worker-ncn-when-iscsi_sbps-group-exists)
* [After initial CSM 1.7 install or upgrade](#after-initial-csm-17-install-or-upgrade)

### CSM install

> During a CSM install, if the administrator wants to enable this feature, they should [create the `iscsi_worker` HSM group](#creating-the-group) as part of [Configure Administrative Access](../../install/configure_administrative_access.md)
> 
> The feature will take effect when [Management Node Personalization](../configuration_management/Management_Node_Personalization.md) runs on the NCNs, after [SAT Bootprep](../system_admin_toolkit/usage/SAT_Bootprep.md) has run.

Reviewer: ```suggestion
During a CSM install, if the administrator wants to enable this feature, they should
[create the `iscsi_worker` HSM group](#creating-the-group) as part of
[Configure Administrative Access](../../install/configure_administrative_access.md).

The feature will take effect when [Management Node Personalization](../configuration_management/Management_Node_Personalization.md)
runs on the NCNs, after [SAT Bootprep](../system_admin_toolkit/usage/SAT_Bootprep.md) has run.
```

    * [Listing group members](#listing-group-members)
    * [Removing a worker from the group](#removing-a-worker-from-the-group)
    * [Deleting the group](#deleting-the-group)
* [Procedures](#procedures)
    * [CSM install](#csm-install)
    * [CSM upgrade from 1.6 to 1.7](#csm-upgrade-from-16-to-17)
    * [Adding or removing worker NCN](#adding-or-removing-worker-ncn)
        * [Adding worker NCN when `iscsi_sbps` group does not exist](#adding-worker-ncn-when-iscsi_sbps-group-does-not-exist)
        * [Adding worker NCN when `iscsi_sbps` group exists](#adding-worker-ncn-when-iscsi_sbps-group-exists)
        * [Removing worker NCN when `iscsi_sbps` group does not exist](#removing-worker-ncn-when-iscsi_sbps-group-does-not-exist)
        * [Removing worker NCN when `iscsi_sbps` group exists](#removing-worker-ncn-when-iscsi_sbps-group-exists)
    * [After initial CSM 1.7 install or upgrade](#after-initial-csm-17-install-or-upgrade)
* [Manual DNS records update](#manual-dns-records-update)

## Overview

The selective iSCSI worker node personalization feature allows administrators to specify which worker NCNs are enabled as

> iSCSI targets. The mechanism by which an administrator specifies this is the `iscsi_worker` [component group](../hardware_state_manager/Component_Groups_and_Partitions.md#groups) in the [Hardware State Manager (HSM)](../../glossary.md#hardware-state-manager-hsm).

Reviewer: ```suggestion
iSCSI targets. The mechanism by which an administrator specifies this is the `iscsi_worker`
[component group](../hardware_state_manager/Component_Groups_and_Partitions.md#groups) in the
[Hardware State Manager (HSM)](../../glossary.md#hardware-state-manager-hsm).
```

beginning of [2. Execute the IUF `management-nodes-rollout` stage](../iuf/workflows/management_rollout.md#2-execute-the-iuf-management-nodes-rollout-stage).

1. Create the `iscsi_worker` HSM group.

   See [Creating the group](#creating-the-group).

1. Perform manual DNS update.

   This step is required any time a worker node is changed from being enabled for iSCSI to being disabled for iSCSI.
   In CSM 1.6, all worker nodes were enabled for iSCSI. Therefore, this step will always be required when first enabling
   this feature on a system that previously was running CSM 1.6.

   See [Manual DNS records update](#manual-dns-records-update).

### Adding or removing worker NCN

When [adding or removing a worker NCN](../node_management/Add_Remove_Replace_NCNs/Add_Remove_Replace_NCNs.md),
additional steps may be required. **Before** performing the add or remove procedure for the worker NCN,
read the section that applies to the current operation and whether or not the `iscsi_sbps` group currently exists:

> > If unsure whether or not the group exists, try [Listing the group members](#listing-group-members).
> > If it succeeds, the group exists (and its member list will be shown). If the group does not exist,
> > the command will fail with an error that it cannot find the group.

Reviewer: ```suggestion
> Notes:
>
> * If unsure whether or not the group exists, try [Listing the group members](#listing-group-members).
>     * If it succeeds, then the group exists (and its member list will be shown).
>     * If the group does not exist, then the command will fail with an error that it cannot find the group.
> * Both before and after the overall add or remove procedure, administrators may wish to validate the iSCSI configuration.
>   For details on how to do this, see [Node Personalization Verification](Node_Personalization_Verification.md).
```

1. Instruct the [Configuration Framework Service (CFS)](../../glossary.md#configuration-framework-service-cfs) to update
   the iSCSI configuration.

   > The latest CSM documentation RPM must be installed on the node where this step is being performed.
   > See [Check for latest documentation](../../update_product_stream/README.md#check-for-latest-documentation).

   (`ncn-mw#`) The `refresh_worker_iscsi_config.py` script is provided to modify the CFS state of the worker NCNs to
   force them to re-run the iSCSI configuration layer. It is run without arguments.

   ```bash
   /usr/share/doc/csm/scripts/operations/configuration/refresh_worker_iscsi_config.py
   ```

   Post personalization verification [Node Personalization Verification](Node_Personalization_Verification.md)is optional if the above script is run.

1. Manually update DNS records, if needed.

    If the changes made in the first step result in a worker being disabled as an iSCSI target when previously
    it had been enabled as an iSCSI target, then an additional manual procedure is required.
    See [Manual DNS records update](#manual-dns-records-update).

> 

Reviewer: ```suggestion
1. Manually update DNS records, if needed.

    If the changes made in the first step result in a worker being disabled as an iSCSI target when previously
    it had been enabled as an iSCSI target, then an additional manual procedure is required.
    See [Manual DNS records update](#manual-dns-records-update).

1. *Optional:* Verify the iSCSI configuration.

    After the CFS node personalization is complete for all of the worker NCNs, administrators may wish to verify the iSCSI configuration.
    See [Node Personalization Verification](Node_Personalization_Verification.md).
```

The [Group commands](#group-commands) section goes over the commands used to create, remove, or modify the group.
The [Procedures](#procedures) section discusses the different contexts in which this may be done, and any additional
steps that must be performed to effectuate the changes.

## Group commands

This section contains example commands for managing selective worker node personalization for iSCSI SBPS.

For more in-depth information on managing HSM groups in general, see
[Manage Component Groups](../hardware_state_manager/Manage_Component_Groups.md).

* [Creating the group](#creating-the-group)
* [Adding a worker to the group](#adding-a-worker-to-the-group)
* [Listing group members](#listing-group-members)
* [Removing a worker from the group](#removing-a-worker-from-the-group)
* [Deleting the group](#deleting-the-group)

### Creating the group

(`ncn-mw#`) Create the `iscsi_worker` group.

```bash
cray hsm groups create --label iscsi_worker --description "iscsi node personalization" --members-ids x3000c0s5b0n0,x3001c0s35b0n0
```

Example output:

```toml
[[results]]
URI = "/hsm/v2/groups/iscsi_worker"
```

> The output does not contain the list of members in the group. If an administrator wishes to confirm that the group was created with the correct members, see [Listing group members](#listing-group-members).

Reviewer: ```suggestion
The output does not contain the list of members in the group. If an administrator wishes to confirm that the group was
created with the correct members, see [Listing group members](#listing-group-members).
```

   the iSCSI configuration.

   > The latest CSM documentation RPM must be installed on the node where this step is being performed.
   > See [Check for latest documentation](../../update_product_stream/README.md#check-for-latest-documentation).

   (`ncn-mw#`) The `refresh_worker_iscsi_config.py` script is provided to modify the CFS state of the worker NCNs to
   force them to re-run the iSCSI configuration layer. It is run without arguments.

   ```bash
   /usr/share/doc/csm/scripts/operations/configuration/refresh_worker_iscsi_config.py
   ```

   Post personalization verification [Node Personalization Verification](Node_Personalization_Verification.md)is optional if the above script is run.

1. Manually update DNS records, if needed.

    If the changes made in the first step result in a worker being disabled as an iSCSI target when previously
    it had been enabled as an iSCSI target, then an additional manual procedure is required.
    See [Manual DNS records update](#manual-dns-records-update).

## Manual DNS records update

One of the things that the iSCSI Ansible playbook does is to add DNS entries for the workers that are enabled for iSCSI.

> However, this playbook does not have logic to remove these entries, in the case where a worker had previously been enabled for iSCSI but now is not. In this case, the administrator must manually remove these entries.
> 
> These entries are added by the `csm.sbps.dns_srv_records` Ansible role in the `csm-config-management` repository in the [Version Control Service (VCS)](../../glossary.md#version-control-service-vcs).

Reviewer: ```suggestion
However, this playbook does not have logic to remove these entries, in the case where a worker had previously been
enabled for iSCSI but now is not. In this case, the administrator must manually remove these entries.

These entries are added by the `csm.sbps.dns_srv_records` Ansible role in the `csm-config-management` repository in
the [Version Control Service (VCS)](../../glossary.md#version-control-service-vcs).
```

1. Perform manual DNS update.

   This step is required any time a worker node is changed from being enabled for iSCSI to being disabled for iSCSI.
   In CSM 1.6, all worker nodes were enabled for iSCSI. Therefore, this step will always be required when first enabling
   this feature on a system that previously was running CSM 1.6.

   See [Manual DNS records update](#manual-dns-records-update).

### Adding or removing worker NCN

When [adding or removing a worker NCN](../node_management/Add_Remove_Replace_NCNs/Add_Remove_Replace_NCNs.md),
additional steps may be required. **Before** performing the add or remove procedure for the worker NCN,
read the section that applies to the current operation and whether or not the `iscsi_sbps` group currently exists:

> Notes:
>
> * If unsure whether or not the group exists, try [Listing the group members](#listing-group-members).
>     * If it succeeds, then the group exists (and its member list will be shown).
>     * If the group does not exist, then the command will fail with an error that it cannot find the group.
> * Both before and after the overall add or remove procedure, administrators may wish to validate the iSCSI configuration.

> >   by running GOSS tests for iSCSI [GOSS tests for SBPS](https://github.com/Cray-HPE/sbps-marshal/blob/main/GOSS_tests_for_sbps.md)

Reviewer: ```suggestion
>   by running [GOSS tests for SBPS](https://github.com/Cray-HPE/sbps-marshal/blob/main/GOSS_tests_for_sbps.md).
```

1. Instruct the [Configuration Framework Service (CFS)](../../glossary.md#configuration-framework-service-cfs) to update
   the iSCSI configuration.

   > The latest CSM documentation RPM must be installed on the node where this step is being performed.
   > See [Check for latest documentation](../../update_product_stream/README.md#check-for-latest-documentation).

   (`ncn-mw#`) The `refresh_worker_iscsi_config.py` script is provided to modify the CFS state of the worker NCNs to
   force them to re-run the iSCSI configuration layer. It is run without arguments.

   ```bash
   /usr/share/doc/csm/scripts/operations/configuration/refresh_worker_iscsi_config.py
   ```

1. Manually update DNS records, if needed.

    If the changes made in the first step result in a worker being disabled as an iSCSI target when previously
    it had been enabled as an iSCSI target, then an additional manual procedure is required.
    See [Manual DNS records update](#manual-dns-records-update).

1. *Optional:* Verify the iSCSI configuration.

    After the CFS node personalization is complete for all of the worker NCNs, administrators may wish to verify the iSCSI configuration.

>     See [Goss tests for iSCSI](https://github.com/Cray-HPE/sbps-marshal/blob/main/GOSS_tests_for_sbps.md).

Reviewer: ```suggestion
    See [Goss tests for SBPS](https://github.com/Cray-HPE/sbps-marshal/blob/main/GOSS_tests_for_sbps.md).
```

> # Managing selective node personalization

Reviewer: The main header (but not the sub-headers) should have every word capitalized, except for ones like `the`, `or`, `for`, `of`, etc. The file should also be renamed to match this.
```suggestion
# Managing Selective Node Personalization
```

> Example command(s) to manage selective worker node personalization for iSCSI SBPS:

Reviewer: The HPE docs style guide says to avoid using `(s)` at the end of words. Just make it plural. If it is really important to highlight that it may be one or it may be more than one, you can say `command or commands`. But in this case, I think just making it plural makes the most sense.
```suggestion
Example commands to manage selective worker node personalization for iSCSI SBPS:
```
